aten/src/ATen/core/dispatch/Dispatcher.h

```diff
@@ -402,9 +402,70 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
 };
 
 namespace detail {
-template<class... Args> inline void unused_arg_(const Args&...) {}
+template <class... Args> inline void unused_arg_(const Args&...) {}
+
+// ReturnValue is intended to capture return values from Dispatcher unboxed
+// kernel calls.
+// A record function may request to get outputs from the kernel calls. For
+// boxed kernels, it's straightforward, the returned values are in the stack
+// object. The stack can be passed to record functions. For unboxed kernels, we
+// need to handle different kinds of return values, cache them temporarily, then
+// release the values for the actual function call return.
+template <typename ReturnType>
+struct ReturnValue {
+  template <typename F, typename... Args>
+  ReturnValue(
+      F&& kernel,
+      const TypedOperatorHandle<ReturnType(Args...)>& op,
+      DispatchKeySet& dispatchKeySet,
+      Args&&... args)
+      : output_{kernel.template call<ReturnType, Args...>(
+            op,
+            dispatchKeySet,
+            std::forward<Args>(args)...)} {}
+  // Wrap the return values in Stack.
+  Stack getOutputs() {
+    Stack stack;
+    impl::push_outputs<ReturnType, false>::copy(output_, &stack);
+    return stack;
+  }
+  // Since we are returning the value, we don't expect the value to be used
+  // afterward. Copy elision and RVO do not apply to class data members. Using
+  // move semantic to avoid copies when possible.
+  ReturnType release() && {
+    return std::move(output_);
+  }
+
+ private:
+  ReturnType output_;
+};
+
+// Handle the lvalue reference differently since it should not be moved.
+template <>
+inline at::Tensor& ReturnValue<at::Tensor&>::release() && {
```

Comment:
This shouldn't be needed. The generic implementation above

```
ReturnType relese() && {
  return std::move(output_);
}
``

will not move lvalues. Calling `std::move` doesn't actually move, it only creates an rvalue reference to the object. And if `ReturnValue` is an lvalue reference, then that rvalue reference gets converted back into an lvalue reference on return and no harm done.

Reply:
Without this specialization, we'd get a compile error: 
```
In file included from buck-out/dev/gen/caffe2/aten/gen_aten=core/TensorMethods.cpp/core/TensorMethods.cpp:9:
caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:437:12: error: non-const lvalue reference to type 'at::Tensor' cannot bind to a temporary of type 'typename std::remove_reference<Tensor &>::type' (aka 'at::Tensor')
    return std::move(output_);
```